vault/core.go

```diff
@@ -1249,110 +1262,137 @@ func (c *Core) unsealPart(ctx context.Context, seal Seal, key []byte, useRecover
 			}
 			masterKey = storedKeys[0]
 		}
-		// Unseal the barrier so we can rekey
-		if err := c.barrier.Unseal(ctx, masterKey); err != nil {
-			return nil, errwrap.Wrapf("error unsealing barrier with constructed master key: {{err}}", err)
-		}
-		defer c.barrier.Seal()
-
-		switch {
-		case c.migrationSeal.RecoveryKeySupported() && c.seal.RecoveryKeySupported():
-			// Set the recovery and barrier keys to be the same.
-			recoveryKey, err := c.migrationSeal.RecoveryKey(ctx)
-			if err != nil {
-				return nil, errwrap.Wrapf("error getting recovery key to set on new seal: {{err}}", err)
-			}
-
-			if err := c.seal.SetRecoveryKey(ctx, recoveryKey); err != nil {
-				return nil, errwrap.Wrapf("error setting new recovery key information during migrate: {{err}}", err)
-			}
+		c.migrationInfo.masterKey = make([]byte, len(masterKey))
+		copy(c.migrationInfo.masterKey, masterKey)
+		c.migrationInfo.recoveryKey = make([]byte, len(recoveryKey))
+		copy(c.migrationInfo.recoveryKey, recoveryKey)
+	}
 
-			barrierKeys, err := c.migrationSeal.GetStoredKeys(ctx)
-			if err != nil {
-				return nil, errwrap.Wrapf("error getting stored keys to set on new seal: {{err}}", err)
-			}
+	return masterKey, nil
+}
 
-			if err := c.seal.SetStoredKeys(ctx, barrierKeys); err != nil {
-				return nil, errwrap.Wrapf("error setting new barrier key information during migrate: {{err}}", err)
-			}
+func (c *Core) migrateSeal(ctx context.Context) error {
+	if c.migrationInfo == nil {
+		return nil
+	}
 
-		case c.migrationSeal.RecoveryKeySupported():
-			// Auto to Shamir, since recovery key isn't supported on new seal
+	// In case of raft storage, when a follower node assumes leadership, if the
+	// migration has already been performed by the previous leader, skip migrating
+	// again.
```

Comment:
Don't we want to do this for every storage type?

Reply:
No. We don't want to.
For all other storage types, each node will be brought up one by one. This case will be caught in `adjustCoreForSealMigration` during startup. But for Raft, all the nodes will be brought up before hand to attain quorum. So, `adjustCoreForSealMigration` will set all the nodes to do migration when they reach unsealed state. Only then the check here comes into play.

Reply:
I don't think that's a fair assumption. If you're using consul storage all the nodes can be brought up at the same time, all can run adjustCoreForSealMigration, but only one will become active and actually do the migration. So when the active node fails over the new active node will need to know not to run a migration. Same goes for all HA storage types i believe.

Reply:
Done.